sqldelight-runtime/src/test/java/com/squareup/sqldelight/SqlDelightDatabaseTest.kt

```diff
@@ -1,76 +1,101 @@
+  @After fun teardown() {
+    database.close()
+  }
 
-  @Test fun afterTransactionRunsAfterSuccessfulTransactionEnds() {
+  @Test fun `afterTransaction runs after transaction commits`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
-      it.markSuccessful()
-      assertThat(counter.get()).isEqualTo(0)
     }
 
     assertThat(counter.get()).isEqualTo(1)
   }
 
-  @Test fun afterTransactionDoesNotRunAfterUnsuccessfulTransactionEnds() {
+  @Test fun `afterTransaction does not run after transaction rollbacks`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
+      rollback()
     }
 
     assertThat(counter.get()).isEqualTo(0)
   }
 
-  @Test fun afterTransactionRunsImmediatelyWithNoTransaction() {
+  @Test fun `afterTransaction runs immediately with no transaction`() {
     val counter = AtomicInteger(0)
     database.afterTransaction { counter.incrementAndGet() }
     assertThat(counter.get()).isEqualTo(1)
   }
 
-  @Test fun afterTransactionRunsAfterParentTransactionEnds() {
+  @Test fun `afterTransaction runs after parent transaction commits`() {
     val counter = AtomicInteger(0)
-    database.newTransaction().use {
+    database.transaction {
       database.afterTransaction { counter.incrementAndGet() }
       assertThat(counter.get()).isEqualTo(0)
 
-      database.newTransaction().use {
+      database.transaction {
         database.afterTransaction { counter.incrementAndGet() }
         assertThat(counter.get()).isEqualTo(0)
-        it.markSuccessful()
-        assertThat(counter.get()).isEqualTo(0)
       }
 
       assertThat(counter.get()).isEqualTo(0)
-      it.markSuccessful()
-      assertThat(counter.get()).isEqualTo(0)
     }
 
     assertThat(counter.get()).isEqualTo(2)
   }
+
+  @Test fun `afterTransaction does not run in nested transaction when parent rolls back`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      database.afterTransaction { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+
+      database.transaction {
+        database.afterTransaction { counter.incrementAndGet() }
+      }
+
+      rollback()
+    }
+
+    assertThat(counter.get()).isEqualTo(0)
+  }
+
+  @Test fun `afterTransaction does not run in nested transaction when nested rolls back`() {
+    val counter = AtomicInteger(0)
+    database.transaction {
+      database.afterTransaction { counter.incrementAndGet() }
+      assertThat(counter.get()).isEqualTo(0)
+
+      database.transaction {
+        database.afterTransaction { counter.incrementAndGet() }
+        rollback()
+      }
+
+      throw AssertionError()
```

Comment:
I still need https://github.com/square/sqldelight/pull/741/files#diff-3640a303ab0bb66b6a07db4bae1aded3R57 for this case since rolling back in a nested transaction should also short-circuit the parent transaction. Otherwise using a `finally` block seems to work fine. I should add some tests for non rollback runtime exceptions though